cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snap-confine-args.h"
+
+#include <string.h>
+
+#include "utils.h"
+
+struct sc_args {
+	// The security tag that the application is intended to run with
+	char *security_tag;
+	// The executable that should be invoked
+	char *executable;
+
+	// Flag indicating that --version was passed on command line.
+	bool is_version_query;
+	// Flag indicating that --classic was passed on command line.
+	bool is_classic_confinement;
+};
+
+struct sc_args *sc_nonfatal_parse_args(int *argcp, char ***argvp,
+				       struct sc_error **errorp)
+{
+	struct sc_args *args = NULL;
+	struct sc_error *err = NULL;
+
+	if (argcp == NULL || argvp == NULL) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc or argv are NULL");
+		goto out;
+	}
+	// Use dereferenced versions of argcp and argvp for convenience.
+	int argc = *argcp;
+	char **const argv = *argvp;
+
+	if (argc == 0) {
+		err = sc_error_init(SC_ARGS_DOMAIN, 0,
+				    "cannot parse arguments, argc is zero");
+		goto out;
+	}
+
+	args = calloc(1, sizeof *args);
+	if (args == NULL) {
+		die("cannot allocate memory for command line arguments object");
+	}
+	// Check if we're being called through the ubuntu-core-launcher symlink.
+	// When this happens we want to skip the first positional argument as it is
+	// the security tag repeated (legacy).
+	bool ignore_first_tag = false;
+	char *basename = strrchr(argv[0], '/');
```

Comment:
Are you sure? I think I check for argc size above. https://github.com/snapcore/snapd/pull/2416/files/19c093cfbbb28775ef98801038fb89092d1dedb4#diff-5c0eca39cb68ec44020b045836523bbaR51